### PR TITLE
Add support for disallowing namespaces

### PR DESCRIPTION
**Pull request by ruudk**
See #50 

Currently it only triggers on `FullyQualified` nodes. 

```neon
    disallowedNamespaces:
        -
            namespace: 'Symfony\Component\HttpFoundation\RequestStack'
            message: 'pass Request via controller instead'
            allowIn:
                - tests/*
        -
            namespace: 'Assert\*'
            message: 'use Webmozart\Assert instead'
```

### CODE REVIEW

**Comment by ruudk** [comment]
You could think: Why not just use `disallowedMethodCalls` and use a wildcard like `Symfony\Component\HttpFoundation\RequestStack::*'`. 

That would work, but only for method calls.

In my case I wanted to prevent the use of `RequestStack` like this:
```php
class MyClass
{
    public function __construct(RequestStack $requestStack) {
        // ..
    }
}
```

**Comment by spaze** [comment]
Thanks. I'm fine with a new config key. Can you please squash the commits or somehow fix the commits so I can review? The "Add tests" commit now changes other areas too, guess it shouldn't.

**Comment by ruudk** [comment]
I'm not with my computer anymore. Please review from GH diff view. Before we merge it I'll squash it.

File: src/NamespaceUsages.php [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+	{
+		if ($node instanceof FullyQualified) {
+			$namespace = $node->toString();
+		} elseif ($node instanceof Node\Stmt\UseUse) {
+			$namespace = $node->name->toString();
+		} elseif ($node instanceof Node\Expr\StaticCall && $node->class instanceof Node\Name) {
+			$namespace = $node->class->toString();
+		} else {
+			return [];
+		}
```
Comment (ruudk) [inline src/NamespaceUsages.php:61]: Can you think of more types that are missing?

File: .gitignore [hunk 2]
```diff
@@ -1,4 +1,4 @@
 /composer.lock
 /vendor
 /tests/tmp
-/tests/.phpunit.result.cache
```
Comment (spaze) [inline .gitignore:4]: Why this change?

File: src/DisallowedNamespace.php [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+	/** @var string|null */
+	private $message;
+
+	/** @var string[] */
+	private $allowIn;
+
+
+	/**
+	 * DisallowedNamespace constructor.
```
Comment (spaze) [inline src/DisallowedNamespace.php:20]: Remove this redundant line.

File: src/DisallowedNamespaceHelper.php [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+	}
+
+
+	private function matchesNamespace(string $pattern, string $value): bool
+	{
+		if ($pattern === $value) {
+			return true;
+		}
+
+		if (fnmatch($pattern, $value, FNM_NOESCAPE)) {
```
Comment (spaze) [inline src/DisallowedNamespaceHelper.php:95]: Why `fnmatch`? Also see my comment in #53

File: src/NamespaceUsages.php (hunk 1 above)
Comment (spaze) [reply src/NamespaceUsages.php:61]: There's a list of things that should be supported in #50, please add tests for all of them.

File: tests/NamespaceUsagesTest.php [hunk 5]
```diff
@@ -0,0 +1,63 @@
+<?php
+declare(strict_types = 1);
+
+namespace Spaze\PHPStan\Rules\Disallowed;
+
+use PHPStan\File\FileHelper;
+use PHPStan\Rules\Rule;
+use PHPStan\Testing\RuleTestCase;
+
+final class NamespaceUsagesTest extends RuleTestCase
```
Comment (spaze) [inline tests/NamespaceUsagesTest.php:10]: Remove `final`

File: tests/NamespaceUsagesTest.php [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+{
+
+	protected function getRule(): Rule
+	{
+		return new NamespaceUsages(
+			new DisallowedNamespaceHelper(new FileHelper(__DIR__)),
+			[
+				[
+					'namespace' => 'Inheritance\Sub',
```
Comment (spaze) [inline tests/NamespaceUsagesTest.php:19]: Add more tests, also for disallowed "parent" namespace, `Inheritance` in this case. Feel free to create new classes, if needed.

File: tests/NamespaceUsagesTest.php [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+					],
+				],
+			]
+		);
+	}
+
+
+	public function testRule(): void
+	{
+		// Based on the configuration above, in this file:
```
Comment (spaze) [inline tests/NamespaceUsagesTest.php:41]: Add the other comment lines to better explain what's expected, see the other test files.

File: src/NamespaceUsages.php [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+	 * @return string[]
+	 * @throws ClassNotFoundException
+	 */
+	public function processNode(Node $node, Scope $scope): array
+	{
+		if ($node instanceof FullyQualified) {
+			$namespace = $node->toString();
+		} elseif ($node instanceof Node\Stmt\UseUse) {
+			$namespace = $node->name->toString();
+		} elseif ($node instanceof Node\Expr\StaticCall && $node->class instanceof Node\Name) {
```
Comment (spaze) [inline src/NamespaceUsages.php:57]: Import the classes used (`use Node\Expr\StaticCall` etc.)

File: src/NamespaceUsages.php [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+namespace Spaze\PHPStan\Rules\Disallowed;
+
+use PhpParser\Node;
+use PhpParser\Node\Name\FullyQualified;
+use PHPStan\Analyser\Scope;
+use PHPStan\Broker\ClassNotFoundException;
+use PHPStan\Rules\Rule;
+
+/**
+ * Reports on usages of classes in general.
```
Comment (spaze) [inline src/NamespaceUsages.php:13]: Remove the comment, it's not correct anyway. Maybe even remove the `@package` line below as well.

**Review by spaze — CHANGES_REQUESTED** [review]
Quick round 1 of review.

Please also squash the commits, thanks to remove unrelated changes.

File: .gitignore (hunk 2 above)
Comment (ruudk) [reply .gitignore:4]: Because every time I run tests it creates `.phpunit.result.cache` in the root directory instead of `tests`. I have to remove it before I commit. Therefore this should be ignored throughout the whole project.

Comment (ruudk) [reply .gitignore:4]: Moved to separate PR for easier review https://github.com/spaze/phpstan-disallowed-calls/pull/54

File: src/NamespaceUsages.php (hunk 1 above)
Comment (ruudk) [reply src/NamespaceUsages.php]: Great, I think I've added all cases now :) 

Except `$this->namespacedObj->call)` (and `$this->x();` that is inherited) as I don't know how to do this. 

Is that really needed?

File: src/DisallowedNamespaceHelper.php (hunk 4 above)
Comment (ruudk) [reply src/DisallowedNamespaceHelper.php:95]: See other PR for explanation :)

File: tests/NamespaceUsagesTest.php (hunk 6 above)
Comment (ruudk) [reply tests/NamespaceUsagesTest.php:35]: Done

**Comment by ruudk** [comment]
Updated the PR. Much better now.

**Review by spaze — APPROVED** [review]

**Comment by spaze** [comment]
Thank you, much better, agree 👍 Made minor changes (some whitespace removal, added namespaces to the docs, removed one `@throws`) directly and merged. Will probably release right away.

**Comment by spaze** [comment]
Just released this in 1.3.0, thank you for your contribution.

**Comment by ruudk** [comment]
Thanks for merging all the PR's and the kind words in the release notes 💙
